kratos/includes/kratos_components.h

```diff
@@ -93,6 +96,25 @@ class KRATOS_API(KRATOS_CORE) KratosComponents
         msComponents.insert(ValueType(rName , &rComponent));
     }
 
+    /**
+     * @brief Adds the source of a component to the list of loaded components.
+     * @param rSrcName the name of the source who added the component
+     * @param rCmpName the name of the component to add
+     * @throws std::runtime_error if the same object name was already registered in the same origin
+     */
+    static void AddSource(const std::string& rSrcName, const std::string& rCmpName)
+    {
+        // Create the container if its the first component source added
+        if (msComponentsSources.find(rSrcName) == msComponentsSources.end()) {
+            msComponentsSources[rSrcName] = std::unordered_set<std::string>();
+        }
+
+        // Check if a different object was already registered with this name
+        auto it_comp =  msComponentsSources.find(rCmpName);
+        KRATOS_ERROR_IF(it_comp != msComponentsSources.end()) << "An object with name \"" << rCmpName << " has already been registered for " << rSrcName << std::endl;
+        msComponentsSources[rSrcName].insert(rCmpName);
+    }
```

Comment:
this function searches 3 times for `rSrcName` in `msComponentsSources`

Can you please improve this?
E.g. sth like
```c++
auto it_found = msComponentsSources.find(rSrcName);
if (it_found == msComponentsSources.end()) {
    it_found = msComponentsSources.insert(rSrcName, std::unordered_set<std::string>()); // probably wrong function, needs the one that returns the new iterator
}

// use it_found ...

```

Reply:
Oh, wait there is a mistake, the second find should be done over:

```
// Check if a different object was already registered with this name
        auto it_comp =  msComponentsSources[rSrcName].find(rCmpName);
        KRATOS_ERROR_IF(it_comp != msComponentsSources[rSrcName].end()) << "An object with name \"" << rCmpName << " has already been registered for " << rSrcName << std::endl;
        msComponentsSources[rSrcName].insert(rCmpName);
```

Does this make more sense?